### DISCUSSION TITLE
Null check at runtime by default + optional disabling

### DISCUSSION THREAD

**Discussion by SylvainGantois** [dotnet/runtime#104677] [Q&A] [answered]
Currently it is possible to run such code where we assign `null` to a non nullable:

`string notANullableString;`
`string? nullableString = null;`
`notANullableString = nullableString;`
`Console.WriteLine("Everything is fine");`

The reasons I saw online for not throwing exceptions were
- runtime null checks can impair performance
- extra flexibility 

Wouldn't it make sense to enable runtime null check by default (generating an exception when assigning `null` to a non-nullable), and having the option to disable them if they are a burden?

Not only that would save us from writing the boiler plate code to check if a non nullable is `null` (ouch), but also probably make programs safer by reducing the risk of unwanted behaviours.

**Comment by julealgon** [comment]
This should be moved to discussions IMHO.

**Comment by julealgon** [comment, marked as answer]
I'm pretty sure the biggest problem with this is that NRT is optional from the beginning... meaning you can be interacting with a lot of libraries that have it turned off, even if you have it turned on yourself.

This means this kind of runtime check would be unreliable at best.

There are also some scenarios today that don't "respect" NRT annotations, such as some serialization-related flows (i.e. `IOptions` binding), so the enforcement would not be comprehensive enough.

I think there is a lot more work that needs to happen before making this an automatic runtime check is feasible, but I'd also love to have it.
